warp/handlers/signature_request_p2p.go

```diff
@@ -0,0 +1,146 @@
+// (c) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package handlers
+
+import (
+	"context"
+	"time"
+
+	"github.com/ava-labs/avalanchego/codec"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+	"github.com/ava-labs/avalanchego/proto/pb/sdk"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	avalancheWarp "github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	"github.com/ava-labs/subnet-evm/warp"
+	"google.golang.org/protobuf/proto"
+)
+
+var _ p2p.Handler = (*SignatureRequestHandlerP2P)(nil)
+
+const (
+	ErrFailedToParse = iota
+	ErrFailedToGetSig
+	ErrFailedToMarshal
+)
+
+// SignatureRequestHandlerP2P serves warp signature requests using the p2p
+// framework from avalanchego. It is a peer.RequestHandler for
+// message.MessageSignatureRequest.
+type SignatureRequestHandlerP2P struct {
+	backend warp.Backend
+	codec   codec.Manager
+	stats   *handlerStats
+}
+
+func NewSignatureRequestHandlerP2P(backend warp.Backend, codec codec.Manager) *SignatureRequestHandlerP2P {
+	return &SignatureRequestHandlerP2P{
+		backend: backend,
+		codec:   codec,
+		stats:   newStats(),
+	}
+}
+
+func (s *SignatureRequestHandlerP2P) AppRequest(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	deadline time.Time,
+	requestBytes []byte,
+) ([]byte, *common.AppError) {
+	// Per ACP-118, the requestBytes are the serialized form of
+	// sdk.SignatureRequest.
+	req := new(sdk.SignatureRequest)
+	if err := proto.Unmarshal(requestBytes, req); err != nil {
+		return nil, &common.AppError{
+			Code:    ErrFailedToParse,
+			Message: "failed to unmarshal request: " + err.Error(),
+		}
+	}
+
+	var (
+		sig [bls.SignatureLen]byte
+		err error
+	)
+	if len(req.Message) == ids.IDLen {
+		// Treat requests with 32-byte [message] as a block hash.
+		var blockID ids.ID
+		copy(blockID[:], req.Message)
+		sig, err = s.GetBlockSignature(blockID)
+		if err != nil {
+			s.stats.IncBlockSignatureMiss()
+		} else {
+			s.stats.IncBlockSignatureHit()
+		}
+	} else {
+		// Otherwise, treat the request as an unsigned warp message.
+		sig, err = s.GetMessageSignature(req.Message)
+		if err != nil {
+			s.stats.IncMessageSignatureMiss()
+		} else {
+			s.stats.IncMessageSignatureHit()
+		}
```

Comment:
I wonder if we should use our internal codec here to parse `req.Message`. I don't think switching based on length is a reliable way.

On the other side, if we need to parse another data for payload, it will again make these messages specific to per-VM. I think in the long run we should probably define those in AvalancheGo, maybe as protos, so that both relayer and VMs can re-use them.
cc @michaelkaplan13 @StephenButtolph

Reply:
Oh sorry just saw this comment after adding my own, but yeah the `req.Message` will always be warp `UnsignedMessage` so we can parse that and then switch based on the typeIDs in it's payload. (Added the comment [here](https://github.com/ava-labs/subnet-evm/pull/1260/files#r1696960581) before seeing this).